datalad/cmd.py

```diff
@@ -98,7 +98,208 @@ def _cleanup_output(stream, std):
+          processing. The callable may transform it in any way, its output
+          (byte-string) is concatenated and provided as stdout return value.
+          The helper functions 'kill_output' and 'capture_output' are
+          provided to either swallow all output (and not relay it to the
+          parent) or to capture all output and provide it as the return
+          value.
+        proc_stderr : callable, optional
+          Like proc_stdout, but for stderr.
+        stdin : byte stream, optional
+          File descriptor like, used as stdin for the process. Passed
+          verbatim to subprocess.Popen().
+
+        Returns
+        -------
+        stdout, stderr
+          Unicode string with the cumulative standard output and error
+          of the process.
+
+        Raises
+        ------
+        CommandError
+          On execution failure (non-zero exit code) this exception is
+          raised which provides the command (cmd), stdout, stderr,
+          exit code (status), and a message identifying the failed
+          command, as properties.
+        """
+        proc_out = (proc_stdout, proc_stderr)
+        if all(p is None for p in proc_out):
+            proc_out = None
+        try:
+            lgr.log(8, "Start running %r", cmd)
+            process = subprocess.Popen(
+                cmd,
+                # from PY37 onwards
+                #capture_output=proc_output,
+                stdout=subprocess.PIPE if proc_out else None,
+                stderr=subprocess.PIPE if proc_out else None,
+                shell=False,
+                cwd=self.cwd,
+                env=self.env,
+                stdin=stdin,
+                # intermediate reports are never decoded anyways
+                # from PY37 onwards
+                #text=False,
+                universal_newlines=False,
+            )
+        except Exception as e:
+            lgr.log(11, "Failed to start %r%r: %s" %
+                    (cmd,
+                     (" under %r" % self.cwd) if self.cwd else '',
+                     exc_str(e)))
+            raise
+
+        try:
+            out = [b'', b'']
+            last_idx = [0, 0]
+            pout = None
+            # make sure to run this loop at least once, even if the
+            # process is already dead
+            keep_going = True
+            while process.poll() is None or keep_going:
+                try:
+                    # get a chunk of output for the specific period
+                    # of time
+                    pout = process.communicate(
+                        timeout=self._poll_period,
+                    )
+                    # we know the process ended at this point
+                    keep_going = False
+                except subprocess.TimeoutExpired as poll:
+                    # this will always be the full report so far,
+                    # not just an output increment
+                    pout = (poll.stdout, poll.stderr)
```

Comment:
On my end, I don't see output stored on timeouts:

```python
import subprocess as sp

proc = sp.Popen(["sh", "-c", "echo a && sleep 3 && echo b"],
                stdout=sp.PIPE, stderr=sp.PIPE)
try:
    proc.communicate(timeout=1)
except sp.TimeoutExpired as exc:
    print(exc.stdout)

print(proc.communicate())
```

```
None
(b'a\nb\n', b'')
```

This is probably the reason why I don't see progress bars when I clone.  Puzzled.

Reply:
Looking at subprocess.py, it doesn't seems like the `TimeoutExpired` that would be raised by `.communicate()` would fill in `stdout` and `stderr` until 580d2782f7 (bpo-37424: Avoid a hang in subprocess.run timeout output capture [...], 2019-09-11).  Before that commit, `subprocess.run` looks to be the only spot that populates `TimeoutExpired`s `stdout` and `stderr`.

v3.9.0a1 is the first tag that contains that commit.  That fits with what I'm seeing (I have 3.7.3), but doesn't make sense with the output that you and Yarik are seeing because I doubt you both are running a development Python version.

Reply:
That is weird. I have Python 3.7.3 and the tests (that also verify the the increments with which the output is processed pass on my machine. However, it makes much more sense now that the same tests failed on the various CIs. But how is my machine special?

Reply:
> I have Python 3.7.3 and the tests (that also verify the the increments with which the output is processed pass on my machine

All of the test_witless_runner.py tests, and in particular `test_runner_incomplete_capture`, pass on my machine as well.  But those tests are not a reliable indication that `stdout` and `stderr` of `TimeoutExpired` are being populated: the tests pass for me even when I apply the following patch.

```diff
diff --git a/datalad/cmd.py b/datalad/cmd.py
index 74003b6bb..8893d883b 100644
--- a/datalad/cmd.py
+++ b/datalad/cmd.py
@@ -230,6 +230,8 @@ def run(self, cmd, proc_stdout=None, proc_stderr=None, stdin=None):
                     # we know the process ended at this point
                     keep_going = False
                 except subprocess.TimeoutExpired as poll:
+                    assert poll.stdout is None
+                    assert poll.stderr is None
                     # this will always be the full report so far,
                     # not just an output increment
                     pout = (poll.stdout, poll.stderr)
```

Reply:
>  I have Python 3.7.3 [...] But how is my machine special?

I'm having no luck coming up with plausible things that could be going on here.  Hoping that maybe the issue is with my machine and that I'm misreading subprocess.el or the cpython git history, I tried in a `debian:latest` docker container with neurodebian sources and Python 3.7.3.  The behavior I see there matches what I report for my local machine: the snippet above shows `None` for `TimeoutExpired.stdout` and `datalad clone https://github.com/datalad/datalad` doesn't show progress (matching the gif I posted).